### PR TITLE
docs: add performance testing doc

### PR DESCRIPTION
**Pull request by akashsinghal**
- Adds performance testing document
- Contains result from latest perf test
- Contains explanation of results
- Defines performance framework
- Explains how Azure pipeline was implemented

### CODE REVIEW

**Review by susanshi — APPROVED** [review]
This is great, thanks Akash.

**Comment by susanshi** [comment]
Not a part of this PR, just curious, if customer wonders what are the remaining work for EXPERIMENTAL_HA work, which doc should they read?  ( this doc mentions a little bit about distributed cache investigation)

File: docs/reference/performance.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+Ratify’s security guarantees rely on its performance at various cluster scales. This document specifies the current ratify performance capabilities. It presents the most recent performance data findings and explains how the data was collected via an automated pipeline.
+
+>Note: These benchmarks are based on Azure end-to-end testing data. The performance framework applied is extensible and cloud-provider agnostic.
+
+>Last test conducted: September 2023
+
+
+## Supported Limits:
+- Single Ratify Pod: up to 1k pod clusters, where each pod results in 50 supply chain artifacts being verified.
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply docs/reference/performance.md:11]: good question. So GK by default ships with 3 replicas. That seemed to be sufficient all the way up to the 10k pod cluster tests.

File: docs/reference/performance.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+Each scenario is grouped by color. Red boxes indicate threshold violations. 
+
+![](../imgs/ratify_perf_results_sept2023.png)
+
+### Test Scenarios Explained
+
+#### Deployments across ACR Regions (Blue)
+
+The results confirmed the hypothesis that as the ACR region size decreases, the registry throttling increases leading to more Ratify timeouts. As pod count increases, the effects of throttling worsen especially in the XS region for ACR. (Note: the extremely large timeout count for 10k pods in XS region is linked to an issue with the HA setup. More investigation required)
```
Comment (binbin-li) [inline docs/reference/performance.md:102]: I wonder if we need to test different ACR regions with the same size, espeically the case that AKS and ACR are in different/same regions.

File: docs/reference/performance.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+### Measurements
+
+1. Registry Throttling
+    - \# of 429 http responses returned from registry
+    - This indicates supported registry RPS in a particular region.
+1. Ratify pod(s) CPU/memory usage: High Memory consumption is correlated with high RPS load on Ratify.
+1. Approximate max verification request duration: Helpful to understand initial non-cached request durations
+3. P95 verification request duration: Shows distribution of request durations through duration of the test. 
+4. Approximate max mutation request duration: Helpful to understand initial non-cache request durations.
```
Comment (binbin-li) [inline docs/reference/performance.md:86]: Wonder if we could add p99 in the future.

File: docs/reference/performance.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+
+## Supported Limits:
+- Single Ratify Pod: up to 1k pod clusters, where each pod results in 50 supply chain artifacts being verified.
+- HA enabled Ratify Pod with 3 replicas: up to 10k pods
+    - NOTE: For Azure users with large kubernetes clusters greater than 200 pods, a medium or large ACR region MUST be used to avoid large ACR throttling.
+
+## Major Findings:
+- Ratify is primarily **constrained by the memory consumption limits (512Mib) and the registry throttling limits** (varies by registry and region for ACR)
+- **Single instance Ratify can support up to 1k pods comfortably.** Workloads larger than that trigger high memory usage. This eventually leads to the pod being Out-of-Memory (OOM) killed.
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply docs/reference/performance.md:17]: possibly. This round of testing did not focus on that metric as I wanted to be very conservative and I think a 50 artifact workload is likely a lot more than a normal image would have.

File: docs/reference/performance.md (hunk 2 above)
Comment (akashsinghal) [reply docs/reference/performance.md:102]: yeah it's definitely a valid point considering latency adds up with the number of serial requests we have. I think in general guidance is that resources should in the same region/close which is why the testing focused on this.

File: docs/reference/performance.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+- **Ratify's memory usage is directly correlated to the number of resources on the cluster.** Memory usage is a good indicator for Ratify deployment scale out. We may consider adding a Horizontal Pod Autoscaler (HPA) based on memory consumption.
+- **Ratify's CPU usage is fairly minimal**. Although CPU usage was not measured over the course of the test, but instead measured once after, the CPU limits seem to be well below the currently limits set. We may consider lowering the CPU request.  
+- **The performance benefit of the distributed cache for high availability usage is inconclusive.** Large cluster tests with distributed cache enabled in an XS region resulted in anomalous behavior and the logs show very slow request processing. We need to investigate why there is a bottleneck compared to when distributed cache is disabled.
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply docs/reference/performance.md:22]: good point. I will create one.

File: docs/reference/performance.md (hunk 3 above)
Comment (akashsinghal) [reply docs/reference/performance.md:86]: Yes that's a good point. I actually did consider p99 however p99 is not as different as p95. With large scale testing, the request # is so large that the distribution of latencies is extremely narrow. P95 vs P99 is almost the same. Initial verification latency on first request magnitudes larger than rest of the requests that we can only really capture that via a max metric.

File: docs/reference/performance.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+## Major Findings:
+- Ratify is primarily **constrained by the memory consumption limits (512Mib) and the registry throttling limits** (varies by registry and region for ACR)
+- **Single instance Ratify can support up to 1k pods comfortably.** Workloads larger than that trigger high memory usage. This eventually leads to the pod being Out-of-Memory (OOM) killed. 
+- **Deployment and Job workloads result in similar Ratify performance.** Concurrent jobs with many replicas seem to yield slightly larger concurrent requests to Ratify. However, the Ratify request durations and resource consumption is similar. 
+- **Minimal registry request throttling (429 reponse) can lead to drastic performance degradation for Ratify.** In large clusters (10k pods+) in XS ACR regions, Ratify's 429 rate spikes. This leads to back-off retries that prolong each verification operation. Longer running requests can lead to more concurrent requests being processed at once, which eventually increases pod memory usage. Throttling also increases the number of request timeouts leading to the K8s workload resource attempting to recreate the pod. This in turn generates more requests to Ratify.
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply docs/reference/performance.md:19]: I think a small # of retries is beneficial. If we don't retry at all, then we will increase the number of verification requests as the workload controller attempts to create new resources. A few retries in close succession can lead to some requests eventually completely within the first tries.

File: docs/reference/performance.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (akashsinghal) [reply docs/reference/performance.md:11]: To follow up on this, GK folks have done internal testing up to 20k pods and they advise that should be sufficient too.

**Comment by akashsinghal** [comment]
@binbin-li @susanshi thanks for the review. I've made the updates that I think we can address in this PR. PTAL

**Review by binbin-li — APPROVED** [review]

**Pull request merged**